### PR TITLE
Ability to specify different Envelope-Sender and SMTP-From address 

### PR DESCRIPTION
**Pull request by ChessSpider**
see https://github.com/gophish/gophish/issues/858 for discussion

Would like to mention; this is the first time I programmed in GoLang, and, I'd like to thank my employer www.podictive.com for enabling me to spend time on this. And you guys to develop this stuff ^^

With that out of the way; this pull request adds an "Envelope-Sender" entry to the Template thingy. If left empty, it will fallback to the SMTP-From in the Sender-settings. This can be used to pass SPF-checks but still send a spoofing email. 

This pull request falls together with a new pull request for Gomail.

### CODE REVIEW

**Comment by ChessSpider** [comment]
right, so it fails due to depending on the non-updated version of gomail.
Gomail fails due to to tests failing due to changed method signature.

Please let me know how you'd like this fixed. I assume reverting the method signature of gomail to what it was and add a new method in gomail for this feature.  But please advise first.

**Comment by ChessSpider** [comment]
oh and I still have to create the mysql db migration.. argh. Ah well, tomorrow

File: models/template.go
Comment (jordan-wright) [inline models/template.go]: If possible, could you `gofmt` the code in this pull request? This ensures consistent formatting across Gophish and is the recommended way to format Golang.

The easiest way to do this is to configure your editor to run a tool like `goimports` on every save.

**Review by jordan-wright — CHANGES_REQUESTED** [review]
I'll leave the rest of the review in the actual comments on the pull request 😄

**Comment by jordan-wright** [comment]
Hey @ChessSpider - thanks so much for sending this over!

If possible, a MySQL migration would be great. In terms of moving forward, I think the best course of action would be for me to review and merge the `gomail` pull request first, and then restart the Travis build here.

Speaking of those builds, could you do me a favor and add a test or two to indicate that this is working properly from the Golang side of things? This would be useful both in Gophish and `gomail`.

I've added a suggestion about formatting the code - this will help so that the changes aren't overwritten with different spacing the next time I save each file 😄 

And finally, you might consider removing the minified JS/CSS that wasn't affected by this change. It's possible it's only whitespace changes- I haven't looked too closely yet. Either way, to keep things consistent with how I've done them in the past, I'll merge this into a branch when things look good and rebuild the minified files myself to ensure they match up with the latest source files.

Thanks again for all your hard work here. It's appreciated both from you and your employer! 🎊

**Comment by ChessSpider** [comment]
Hi @jordan-wright just a small reminder: i think im all finished now. 
When you're able, you can check https://github.com/gophish/gomail/pull/2#issuecomment-371530088 and this commit and hopefully merge it too!

**Comment by jordan-wright** [comment]
Hey @ChessSpider, just wanted to send an update. Things have been busy here so I haven't had time to get this merged into a branch yet. I should be able to knock this out in the next day or two.

Thanks again for sending this over. I'll keep you updated! 😄

**Comment by ChessSpider** [comment]
now all tests are OK except on Go 1.7. Don't know why, as it works on 1.8 and 1.9 and it doesn't appear to crash on any of the code that I've written.

When do you expect time to look at this? I  kinda need this for a project and we really only use Stable official releases in production.

**Comment by jordan-wright** [comment]
I think the test failures are due to the postgres dependency adding a single line that uses `time.Until` which is a function added in Go 1.8.

Oh well...

These are looking good! I'm going to go ahead and merge the changes into a branch so I can review things further and help take this over the finish line. As soon as that's done (hopefully tomorrow night or Wednesday) I'll tag a new release with the changes 😄 

Thanks again for all your awesome work here!

**Review by jordan-wright — APPROVED** [review]

**Comment by jordan-wright** [comment]
Hey @ChessSpider just wanted to check in on this and give an update. I got a chance to dive a bit more into the code, and there's some changes I'll likely make so I wanted to keep you in the loop.

Specifically:

* I think the biggest change is that I'll probably move the envelope sender setting into the Sending Profile. I know the comment in the issue suggesting not wanting to change this per campaign, but the same inconvenience is there if it's in a template (ie you'd have to make a new template for every campaign). When I look at it now, there's a separation that's hard to keep track of since we default the value to something that's in the Sending Profile, so I think keeping them grouped together will make it a bit cleaner.

* I think the tooltips might be backwards. Right now, the SMTP From in the Sending Profile suggests setting it to something that bypasses SPF checks, but SPF [checks the envelope sender, not the From: header](http://www.openspf.org/FAQ/Envelope_from_scope), and vice-versa with the envelope from tooltip. I'll probably make these tooltips more vague in general to prevent confusion. More information most of the time means more support tickets with questions about what the terms mean 😉 

I think those two pretty much cover it. Thanks to your awesome work, all the plumbing is there so it should be a quick and easy migration + cleanup. Thanks again for all your hard work - I'll keep you updated!

**Comment by ChessSpider** [comment]
Hi Jordan. When testing please **rebuild the minified js** files otherwise it doesn't work! I lost some time on this, thinking I was going crazy.

About SPF, my description is correct. Please check the following phishing email i sent to my gmail account using the gophish platform with my changes:

spf-pass


with the following headers:

```
Delivered-To: machiel--------------@gmail.com
Received: by 10.192.148.140 with SMTP id d12csp3595903imc;
        Tue, 10 Apr 2018 01:38:27 -0700 (PDT)
X-Google-Smtp-Source: AIpwx4+YDmJbvpVD0i0on6faq1wn6MSbojvGwa4D7HOMTD5sco84hAqKjUm9IBkE6COxn8iTfEkg
X-Received: by 10.80.173.75 with SMTP id z11mr1919344edc.306.1523349507069;
        Tue, 10 Apr 2018 01:38:27 -0700 (PDT)
ARC-Seal: i=1; a=rsa-sha256; t=1523349507; cv=none;
        d=google.com; s=arc-20160816;
        b=U3UPRhx4qZNdbrIH/exAukDc22uKYsyReI//zbOsI7tn+ySwuOWOGI7g3Y6rfX4VZz
         KgDEM7F9Mf4TN4GObC/qAOCeo8xgKTBNOMgCZ6lgKu3uh+TNwi0ZWpPaJ46+uAa7HO5A
         6+38zLwaWI3xwpnYLOrqKH3NtQWy1yCI4UWU79BGp8wRbrUXeqdoCbtZD9rhvsQeeffx
         3fmb17NTpQyCeF5/p8KrGX+Irk0xBgBcz3jYxTSpQAl8ZNXXeCAvpEartiMG801A/TNm
         v8TnxvydIdDHggwG/e7OwHlo3RtfCodR1ve2BaySla2kZ0O2ghgjvJJUKqKJBPHyywVP
         imCw==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=google.com; s=arc-20160816;
        h=content-transfer-encoding:to:subject:from:date:mime-version
         :message-id:arc-authentication-results;
        bh=g3zLYH4xKxcPrHOD18z9YfpQcnk/GaJedfustWU5uGs=;
        b=0pK0qXtU7XQGVovjU4FxLh0M7qY0LbsFWWq/4nOJHDF7eXbMbjOEDwnbXIeqIudU6v
         FF3fPmVFfBzo5nkDXaIGch9CBan2Cj3df/skVEoxIAAH0GYSYgKcsXnIAUhaB3704RyB
         4TDKzS2GKwsZlgOERsA2WOQZ5e/LRUsBWyy2xjtX/ZERM5HdEBIwzPAl5Yfz7KzTNdck
         huMQHNnP04D+9z2g/jUrXR2e7c3ayuSQu4tNWvmdNKQAlsklu67dCuXBXUSyoS4yVhSU
         0voEvzamvIlF63z5Fwmto+1boeMSSHMws911veBXWgdgdoq9dmBsxB6X3BKsicmHlS2+
         PCxg==
ARC-Authentication-Results: i=1; mx.google.com;
       spf=pass (google.com: domain of noreply@delxxxxxxxxxxxxxxx.xxxtld designates 136.144.x.x as permitted sender) smtp.mailfrom=noreply@delxxxxxxxxxxxxxxx.xxxtld
Return-Path: <noreply@delxxxxxxxxxxxxxxx.xxxtld>  **SMTP FROM**
Received: from delxxxxxxxxxxxxxxx.xxxtld (136-144-x-x.x.hostingcopany.net. [136.144.x.x])
        by mx.google.com with ESMTPS id m48si204875edm.67.2018.04.10.01.38.26
        for <machiel--------------@gmail.com>
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-GCM-SHA256 bits=128/128);
        Tue, 10 Apr 2018 01:38:27 -0700 (PDT)
Received-SPF: pass (google.com: domain of noreply@delxxxxxxxxxxxxxxx.xxxtld designates 136.144.x.x as permitted sender) client-ip=136.144.x.x;
Authentication-Results: mx.google.com;
       spf=pass (google.com: domain of noreply@delxxxxxxxxxxxxxxx.xxxtld designates 136.144.x.x as permitted sender) smtp.mailfrom=noreply@delxxxxxxxxxxxxxxx.xxxtld
Message-Id: <5acc7803.b38b500a.5c2fa.2d37SMTPIN_ADDED_MISSING@mx.google.com>
Mime-Version: 1.0
Date: Tue, 10 Apr 2018 08:38:26 +0000
From: test <justwhat-hoaware@adksdjfsdfk.com>  **SPF FROM**
Subject: asdtest
To: Machiel -- <machiel--------------@gmail.com>
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: quoted-printable
```

**Comment by ChessSpider** [comment]
> I think the biggest change is that I'll probably move the envelope sender setting into the Sending Profile. I know the comment in the issue suggesting not wanting to change this per campaign, but the same inconvenience is there if it's in a template (ie you'd have to make a new template for every campaign). When I look at it now, there's a separation that's hard to keep track of since we default the value to something that's in the Sending Profile, so I think keeping them grouped together will make it a bit cleaner.


please don't; the main useability problem I have with GoPhish at the moment is the amount of screens I have to navigate to create a new campain. With your suggestion, I'd still have to create a duplicate Sender profile with exactly the same settings except for the Envelope Sender.

I put the Envelope Sender in the Email Template on purpose, as it relates to what a victim sees in the email. If you insist on moving it, please move it to the New Campain screen so I don't 

However the way I use GoPhish now is that, per phishing scenario, I have to create a victim and scenario specific Template and Landing-page. Not also having to add a scenario specific sender profile would make me happy :)

**Comment by ChessSpider** [comment]
> Thanks to your awesome work, all the plumbing is there so it should be a quick and easy migration + cleanup. Thanks again for all your hard work - I'll keep you updated!

Thanks ! it was a pleasure ^^. And a bit confusing at times (due to GoLang)

**Comment by ChessSpider** [comment]
For the interested parties
```
FROM golang:stretch
MAINTAINER Machiel Pronk <machiel.pronk@podictive.com>

RUN apt-get update && \
apt-get install --no-install-recommends -y \
git \
unzip \
ca-certificates \
wget golang && \
apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*


RUN go version && go get github.com/gophish/gophish
WORKDIR /go/src/github.com/gophish/gophish
RUN git config user.email "you@example.com" &&  git remote set-url origin https://github.com/chessspider/gophish && git pull
#RUN git checkout 986-custom-envelope-sender
RUN go build

RUN sed -i "s|127.0.0.1|0.0.0.0|g" config.json && \
chmod +x gophish

EXPOSE 3333 80
CMD ["./gophish"]```

**Comment by ChessSpider** [comment]
> I put the Envelope Sender in the Email Template on purpose, as it relates to what a victim sees in the email. If you insist on moving it, please move it to the New Campain screen.
Additionally, I also want to expand the Email Template to allow for custom headers, e.g. setting a Reply-To header. And if feasible with the current setup, a custom To: header as well.

**Comment by jordan-wright** [comment]
Hey there! I think there may be some mixups in the different fields. From your example you're saying:

> Return-Path: noreply@delxxxxxxxxxxxxxxx.xxxtld SMTP FROM
From: test justwhat-hoaware@adksdjfsdfk.com ENVELOPE FROM

When it's actually the opposite. You'll notice this message in Gmail's output:

> Received-SPF: pass (google.com: domain of noreply@delxxxxxxxxxxxxxxx.xxxtld designates 136.144.x.x as permitted sender) client-ip=136.144.x.x;

This matches the expected format of SPF results found in the [RFC](https://tools.ietf.org/html/rfc7208#section-9.2). In this case, `noreply@delxxxxxxxxxxxxxxx.xxxtld` is the envelope (MAIL FROM) address, while `test <justwhat-hoaware@adksdjfsdfk.com>` is the "From" header. As mentioned in the RFC, SPF is based on the former.

That said, in terms of where to put this stuff, I think I'm still going to put the envelope address in the Sending Profile _for now_. In the short term, I want all SMTP settings to be organized around the SMTP settings, keeping the email template strictly to be the content of the email.

I say "for now", because your comments make me think there's probably room to move some SMTP settings to the campaign. We're already going to be putting the scheduling options there, so I can see a valid argument that the From and Envelope Sender could also belong there. But I don't want to hold this off until that happens - I'd rather get this earlier, keep it logically grouped for the short-term, and move things as part of the larger campaign builder in #223.

Hope this helps explain the reasoning! As always, thanks for the great work, and the great discussion.

**Comment by adambirds** [comment]
Glad to see this is coming along. And I definitely agree with @ChessSpider, it would be better having the setting on the template as it is directly linked to the email you are sending, and we wouldn't need another sending profile.

**Comment by dsmurf** [comment]
> For the interested parties
> 
> ```
> FROM golang:stretch
> 
> RUN apt-get update && \
> apt-get install --no-install-recommends -y \
> git \
> unzip \
> ca-certificates \
> wget golang && \
> apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
> 
> 
> RUN go version && go get github.com/gophish/gophish
> WORKDIR /go/src/github.com/gophish/gophish
> RUN git config user.email "you@example.com" &&  git remote set-url origin https://github.com/chessspider/gophish && git pull
> #RUN git checkout 986-custom-envelope-sender
> RUN go build
> 
> RUN sed -i "s|127.0.0.1|0.0.0.0|g" config.json && \
> chmod +x gophish
> 
> EXPOSE 3333 80
> CMD ["./gophish"]```
> ```

if I run this script on my GoPhish server will this be enough?

**Comment by ChessSpider** [comment]
Do you know how to run Docker/Dockerfiles? You need to portforward the appropiate ports for example. 

Be aware that my branch is outdated now with gophish master, I have not updated it since +/- april ish

**Comment by ChessSpider** [comment]
Do you know how to run Docker/Dockerfiles? You need to portforward the appropiate ports for example. 

Be aware that my branch is outdated now with gophish master, I have not updated it since +/- april ish

**Comment by kkkkkky** [comment]
In my environment, add a custom header "Sender" to email headers and set it to the real smtp from address, then you can fill anything you like to "From" section in the sending profile.

**Comment by Georg311** [comment]
This hasn't been merged to master, has it?

**Pull request merged**
